File: CODE/debug-read-write-big-data.R
#!/usr/bin/env Rscript
cat("\n\n>>> Running debug-read-write-big-data.R ...\n\n")
suppressWarnings(suppressPackageStartupMessages(library(arrow)))
suppressWarnings(suppressPackageStartupMessages(library(qs2)))
suppressWarnings(suppressPackageStartupMessages(library(tictoc)))
if(!exists("N_QS_THREADS")) {
  if(exists("project_dir")) source(file.path(project_dir, "CODE", "functions-and-definitions.R")) else source("functions-and-definitions.R")
}

cat("\n=======================================================\n")
cat("OPTION A: Legacy Pipeline (Direct List Load via qs2)\n")
cat("=======================================================\n")
cat("  >> Loading 1M DataFrames recursively from qs2...\n")
tic()
list_from_qs <- qs_read("shared_data_from_julia.qs", nthreads = N_QS_THREADS)
toc()

cat("\n=======================================================\n")
cat("OPTION B: Modern Pipeline (Arrow Load -> Reshape in Memory)\n")
cat("=======================================================\n")
cat("  >> Zero-copy parsing the 2.2GB Flattened DataFrame from Arrow...\n")
tic()
df_arrow <- read_feather("shared_data_from_julia.arrow")
toc()

cat("  >> Reshaping Arrow Flattened DataFrame back into a List of 1M DataFrames using split()...\n")
tic()
# Note: Since the ID column in Julia is named 'ID', we use $ID here.
list_from_arrow <- split(df_arrow, df_arrow$ID)
toc()

cat("\nBenchmark Complete!\n")

cat("\n\n>>> debug-read-write-big-data.R successfully finished running ✅\n\n")

File: CODE/debug-read-write-big-data.jl
#!/usr/bin/env julia
println("\n\n>>> Running debug-read-write-big-data.jl ...\n\n")
using RCall, Arrow, DataFrames, Serialization, Random, Printf

println("Generating 1 Million DataFrames inside an Array... (This may take a minute)")
N_individuals = 1_000_000
N_rows_per = 40

list_of_dfs = Vector{DataFrame}(undef, N_individuals)
for i in 1:N_individuals
    list_of_dfs[i] = DataFrame(
        ID = fill(i, N_rows_per),
        time = 1.0:1.0:N_rows_per,
        v1 = rand(N_rows_per),
        v2 = rand(N_rows_per),
        v3 = rand(N_rows_per),
        v4 = rand(N_rows_per),
        v5 = rand(N_rows_per)
    )
end

println("-> DataFrames Array generated. Memory roughly ~2.2 GB.")

println("\n=======================================================")
println("OPTION A: Legacy Pipeline (RCall memory transfer -> qs2)")
println("=======================================================")
println("  >> Timing the transfer of 1M DataFrames to R memory via @rput...")
@time @rput list_of_dfs

println("  >> Timing the save to disk via R's `qs2::qs_save`...")
@time R"qs2::qs_save(list_of_dfs, 'shared_data_from_julia.qs', nthreads = N_QS_THREADS)"


println("\n=======================================================")
println("OPTION B: Modern Pipeline (Julia Flatten -> Arrow native)")
println("=======================================================")
println("  >> Timing the flattening of 1M DataFrames into a single continuous table...")
@time df_flat = reduce(vcat, list_of_dfs)

println("  >> Timing the save to disk natively via `Arrow.write`...")
@time Arrow.write("shared_data_from_julia.arrow", df_flat)

println("\nJulia sequence complete! Run the R script next to benchmark the loading.")

println("\n\n>>> debug-read-write-big-data.jl successfully finished running ✅\n\n")
